src/lib/dom-api-mutation.html

```diff
@@ -0,0 +1,280 @@
+<!--
+@license
+Copyright (c) 2014 The Polymer Project Authors. All rights reserved.
+This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+-->
+<link rel="import" href="settings.html">
+<script>
+(function() {
+  'use strict';
+
+    var DomApi = Polymer.DomApi.ctor;
+    var Settings = Polymer.Settings;
+    var hasDomApi = Polymer.DomApi.hasDomApi;
+
+    DomApi.Mutation = function(domApi) {
+      this.domApi = domApi;
+      this.node = this.domApi.node;
+      this._listeners = [];
+      this._addedNodes = [];
+      this._removedNodes = [];
+    };
+
+    DomApi.Mutation.prototype = {
+
+      addListener: function(callback, options) {
+        options = options || {};
+        this._ensureSetup(options);
+        var listener = {fn: callback, options: options};
+        this._notifyInitial(listener);
+        return this._listeners.push(listener);
+      },
+
+      removeListener: function(handle) {
+        this._listeners.splice(handle - 1, 1);
+        if (!this._hasListeners()) {
+          this._ensureCleanup();
+        }
+      },
+
+      _ensureSetup: function(options) {
+        if (!this._isSetup) {
+          this._isSetup = true;
+          this._observeContentElements(this.domApi.childNodes);
+        }
+      },
+
+      _ensureCleanup: function() {}, // abstract
+
+      _hasListeners: function() {
+        return Boolean(this._listeners.length);
+      },
+
+      addNode: function(node) {
+        if (this._hasListeners()) {
+          this._addedNodes.push(node);
+          this._scheduleNotify();
+        }
+      },
+
+      removeNode: function(node) {
+        if (this._hasListeners()) {
+          this._removedNodes.push(node);
+          this._scheduleNotify();
+        }
+      },
+
+      addAllNodes: function(force) {
+        if (this._hasListeners() || force) {
+          var c$ = this.domApi.childNodes;
+          if (c$.length) {
+            for (var i=0, c; (i < c$.length) && (c=c$[i]); i++) {
+              this._addedNodes.push(c);
+            }
+            this._scheduleNotify();

```

Comment:
Is there a chance for double-notification in case of:

* observe no. 1
* initial distribution (schedules notify because there is a listener)
* observe no. 2 (notifies initial & receives scheduled notification also?)

`_notifyInitial` should be gated on whether initial distribution notification has been sent yet